File: python/inline_sql_helper/detection.py
"""Source-level SQL candidate detection for Python string syntax units."""

from dataclasses import dataclass
from typing import Literal

from inline_sql_helper.literals import SupportedLiteral, UnsupportedLiteral
from inline_sql_helper.positions import SourceMap, SourceSpan


@dataclass(frozen=True, slots=True)
class SqlDetection:
    """One source-level SQL detection result."""

    matched: bool
    marker_span: SourceSpan | None
    sql_span: SourceSpan | None
    reason: Literal["marker", "keyword", "none"]


_MARKERS = frozenset({"-- sql", "--sql"})
_KEYWORDS = (
    "select",
    "with",
    "insert",
    "update",
    "delete",
    "merge",
    "create",
    "alter",
    "drop",
    "truncate",
    "explain",
)
_ASCII_WHITESPACE = frozenset(" \t\r\n")


def _continues_identifier(character: str) -> bool:
    """Return whether *character* continues a Python/Unicode identifier."""
    return bool(character) and (character == "_" or ("A" + character).isidentifier())


def _detect_source_slice(text: str, base: int) -> SqlDetection:
    """Detect an explicit marker or leading SQL keyword in one source slice."""
    cursor = 0
    for line in text.splitlines(keepends=True):
        body = line.rstrip("\r\n")
        if body.strip(" \t") == "":
            cursor += len(line)
            continue
        if body.strip(" \t").casefold() in _MARKERS:
            marker = SourceSpan(base + cursor, base + cursor + len(body))
            return SqlDetection(
                True,
                marker,
                SourceSpan(base + cursor + len(line), base + len(text)),
                "marker",
            )
        break

    significant = 0
    while significant < len(text) and text[significant] in _ASCII_WHITESPACE:
        significant += 1
    folded = text[significant:].casefold()
    for keyword in _KEYWORDS:
        if not folded.startswith(keyword):
            continue
        following = text[significant + len(keyword) : significant + len(keyword) + 1]
        if not _continues_identifier(following):
            return SqlDetection(
                True,
                None,
                SourceSpan(base + significant, base + len(text)),
                "keyword",
            )
    return SqlDetection(False, None, None, "none")


def detect_sql(
    literal: SupportedLiteral | UnsupportedLiteral,
    source_map: SourceMap,
) -> SqlDetection:
    """Detect SQL from physical source characters without evaluating escapes."""
    span = (
        literal.content_span
        if isinstance(literal, SupportedLiteral)
        else literal.detection_content_span
    )
    if span is None:
        return SqlDetection(False, None, None, "none")
    return _detect_source_slice(source_map.slice(span), span.start)

File: src/python-analysis/detection.ts
import { SourceMap, SourceSpan } from "./positions.js";
import type { SupportedLiteral, UnsupportedLiteral } from "./tokenizer.js";

/** One source-level SQL detection result. */
export interface SqlDetection {
  readonly matched: boolean;
  readonly markerSpan: SourceSpan | undefined;
  readonly sqlSpan: SourceSpan | undefined;
  readonly reason: "marker" | "keyword" | "none";
}

const MARKERS = new Set(["-- sql", "--sql"]);
const KEYWORDS = [
  "select",
  "with",
  "insert",
  "update",
  "delete",
  "merge",
  "create",
  "alter",
  "drop",
  "truncate",
  "explain",
];
const ASCII_WHITESPACE = new Set([" ", "\t", "\r", "\n"]);

/** Return whether *character* continues a Python/Unicode identifier. */
function continuesIdentifier(character: string): boolean {
  if (character === "") return false;
  if (character === "_") return true;
  return /[\p{L}\p{N}]/u.test(character);
}

/** Split text into lines, keeping terminators, like Python splitlines(True). */
function splitLinesKeepends(text: string): readonly string[] {
  const lines: string[] = [];
  const parts = text.split(/(\r\n|\r|\n)/);
  for (let index = 0; index < parts.length; index += 2) {
    lines.push(`${parts[index] ?? ""}${parts[index + 1] ?? ""}`);
  }
  return lines;
}

/** Detect an explicit marker or leading SQL keyword in one source slice. */
function detectSourceSlice(text: string, base: number): SqlDetection {
  let cursor = 0;
  for (const line of splitLinesKeepends(text)) {
    const body = line.replace(/[\r\n]+$/, "");
    if (body.trim() === "") {
      cursor += line.length;
      continue;
    }
    if (MARKERS.has(body.trim().toLowerCase())) {
      const marker = new SourceSpan(base + cursor, base + cursor + body.length);
      return {
        matched: true,
        markerSpan: marker,
        sqlSpan: new SourceSpan(base + cursor + line.length, base + text.length),
        reason: "marker",
      };
    }
    break;
  }

  let significant = 0;
  while (significant < text.length && ASCII_WHITESPACE.has(text[significant] ?? "")) {
    significant++;
  }
  const folded = text.slice(significant).toLowerCase();
  for (const keyword of KEYWORDS) {
    if (!folded.startsWith(keyword)) continue;
    const following = text[significant + keyword.length] ?? "";
    if (!continuesIdentifier(following)) {
      return {
        matched: true,
        markerSpan: undefined,
        sqlSpan: new SourceSpan(base + significant, base + text.length),
        reason: "keyword",
      };
    }
  }
  return { matched: false, markerSpan: undefined, sqlSpan: undefined, reason: "none" };
}

/** Detect SQL from physical source characters without evaluating escapes. */
export function detectSql(
  literal: SupportedLiteral | UnsupportedLiteral,
  sourceMap: SourceMap,
): SqlDetection {
  const span = "contentSpan" in literal ? literal.contentSpan : literal.detectionContentSpan;
  if (span === undefined) {
    return { matched: false, markerSpan: undefined, sqlSpan: undefined, reason: "none" };
  }
  return detectSourceSlice(sourceMap.slice(span), span.start);
}
